quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -706,9 +780,189 @@ impl IngestController {
+            .protect_future(self.metastore.open_shards(open_shards_request))
+            .await
+        {
+            Ok(open_shards_response) => open_shards_response,
+            Err(error) => {
+                error!(%error, "failed to rebalance shards");
+                return;
+            }
+        };
+        let init_shards_response = self
+            .init_shards(&open_shards_response.subresponses, progress)
+            .await;
+
+        for init_shard_success in init_shards_response.successes {
+            let shard = init_shard_success.shard().clone();
+            let index_uid = shard.index_uid().clone();
+            let source_id = shard.source_id.clone();
+            model.insert_shards(&index_uid, &source_id, vec![shard.clone()]);
+
+            let source_uid = SourceUid {
+                index_uid,
+                source_id,
+            };
+            // We temporarily disable the ability the scale down the number of shards for the source
+            // to avoid closing the shards we just opened.
+            model.drain_scaling_permits(&source_uid, ScalingMode::Down);
+        }
+        for init_shard_failure in init_shards_response.failures {
+            let shard_id = init_shard_failure.shard_id();
+            shards_to_close.remove(shard_id);
+        }
+        let mut per_leader_shards_to_close: HashMap<String, Vec<ShardPKey>> = HashMap::new();
+
+        for (leader_id, shard_pkey) in shards_to_close.into_values() {
+            per_leader_shards_to_close
+                .entry(leader_id)
+                .or_default()
+                .push(shard_pkey);
+        }
+        let mut close_shards_futures = FuturesUnordered::new();
+
+        for (leader_id, shard_pkeys) in per_leader_shards_to_close {
+            let Some(mut ingester) = self.ingester_pool.get(&leader_id) else {
+                warn!("failed to close shards: ingester `{leader_id}` is unavailable");
+                continue;
+            };
+            let shard_ids = shard_pkeys
+                .iter()
+                .map(|shard_pkey| ShardIds {
+                    index_uid: shard_pkey.index_uid.clone(),
+                    source_id: shard_pkey.source_id.clone(),
+                    shard_ids: shard_pkey.shard_id.clone().into_iter().collect(),
+                })
+                .collect();
+            #[allow(deprecated)]
+            let shards_to_close_request = CloseShardsRequest {
+                shard_pkeys,
+                shard_ids,
+            };
+            let close_shards_future = async move {
+                tokio::time::timeout(
+                    Duration::from_secs(3),
+                    ingester.close_shards(shards_to_close_request),
+                )
+                .await
+            };
+            close_shards_futures.push(close_shards_future);
+        }
+        let mailbox = mailbox.clone();
+        let close_shards_future = async move {
```

Comment:
can we extract that in a different function?

one benefit is that it would make some important detail of the implementation clearer: 
- the shard opening happens sync to the control plane actor, but the closing happens on a separate task.
- `_rebalance_permit` is dropped after closed is done.